addressing final review (#60)
OpenLiberty/draft-guide-reactive-rest-client: README.adoc

:mac: MAC
:win: WINDOWS
:linux: LINUX
= Consuming RESTful services using the reactive JAX-RS client

[.hidden]
NOTE: This repository contains the guide documentation source. To view the guide in published form, view it on the https://openliberty.io/guides/{projectid}.html[Open Liberty website].

Learn how to use a reactive JAX-RS client to invoke RESTful microservices over HTTP asynchronously.

== What you'll learn

> You'll learn how to create a reactive JAX-RS client application using two different JAX-RS reactive providers, Apache CXF and Eclipse Jersey. The JAX-RS client is an API used to communicate with RESTful web services. The API makes it very easy to consume a web service exposed using the HTTP protocol and means that you can efficiently implement client-side applications. The reactive client extension to JAX-RS is an API that enables you to use the reactive programming model when using the JAX-RS client.

Reviewer: ```suggestion
You'll learn how to create a reactive JAX-RS client application using the default JAX-RS reactive provider and a pluggable reactive provider for RxJava published by Eclipse Jersey. The JAX-RS client is an API used to communicate with RESTful web services. The API makes it very easy to consume a web service exposed using the HTTP protocol and means that you can efficiently implement client-side applications. The reactive client extension to JAX-RS is an API that enables you to use the reactive programming model when using the JAX-RS client.
```

Sorry, I missed this one before.

Reactive programming is an extension of asynchronous programming and focuses on the flow of data through data streams. 
Reactive applications process data when it becomes available and responds to requests as soon as the processing is complete; the request and response to the application are decoupled so that the application is not blocked from responding to other requests in the meantime. 
Reactive applications can run faster than synchronous applications, which provides a much smoother user experience.

> The application in this guide demonstrates how the JAX-RS client accesses remote RESTful services using asynchronous method calls. You'll first take a look at the supplied client application which uses CXF, the JAX-RS reactive provider used by default in Open Liberty, then you'll modify the client application to instead use Jersey, an alternative JAX-RS reactive provider. Both CXF and Jersey provide third-party reactive libraries, including RxJava, and have been tested for use in Open Liberty.

Reviewer: ```suggestion
The application in this guide demonstrates how the JAX-RS client accesses remote RESTful services using asynchronous method calls. You'll first take a look at the supplied client application which uses JAX-RS's default `CompletionStage`-based provider, then you'll modify the client application to instead use Jersey's RxJava provider, an alternative JAX-RS reactive provider. Both Jersey and Apache CXF provide third-party reactive libraries for RxJava, and have been tested for use in Open Liberty.
```

A [hotspot=countdownlatch file=1]`CountDownLatch` object is used to keep track of how many asynchronous requests are being waited on. 
The `CountDownLatch` object will countdown towards 0 after each completion of a thread via the [hotspot=countdown1 hotspot=countdown2 file=1]`countdown` method call. 
This means that the value returns only after the thread that's retrieving the value is complete.
The [hotspot=await file=1]`await` method stops and waits until all of the requests have completed. 
While waiting for this countdown to complete, the main thread is free to perform other tasks. 

> In this case, no such task is present, but a simple task can be given to occupy the thread.

Reviewer 2: I know I copied this directly from the original version but I'm not sure now what it means: "a simple task can be given..." - does it mean that the user should do something? We could remove from "but a simple task..." to the end of the sentence. The problem is that it's phrased very passively - it's not clear who, if anyone, should give the thread a simple task to do.

[role='command']
include::{common-includes}/gitclone.adoc[]

> == Creating a web client using the CXF reactive provider

Reviewer: The example shown here is for the default JAX-RS API, not the CXF reactive provider, so we should undo this change.  The way to tell which reactive provider will be used is the `rx(...)` method.   If there are no parameters (as it is in this case), then it is the default JAX-RS Provider - and it will always return a `CompletionStage`. If it is a non-default provider, then there will be a class passed in to the `rx` method (like `rx(RxObservableInvoker.class)` for Jersey or `rx(ObservableRxInvoker.class)` for CXF).

Keep in mind that neither Jersey's nor CXF's provider ships with Liberty - both require that the user package the provider with their app.

Reviewer 2: Ah, sorry, both. That was my interpretation of the notes you put in the other issue.

InventoryClient.java
[source, java, linenums, role='code_column hide_tags=copyright']
----
include::defaultrx/query/src/main/java/io/openliberty/guides/query/client/InventoryClient.java[]
----

Navigate to the `start` directory to begin.

> The CXF reactive provider is provided by default in Open Liberty so that you can write applications using reactive APIs like RxJava. 

Reviewer: This should probably change back to the original text - CXF reactive provider is not provided by default - only the JAX-RS reactive API (which uses `CompletionStage`) is provided by default.  See previous comment.

[role="code_command hotspot file=1", subs="quotes"]
----
#Create the `QueryResource` class.#
`query/src/main/java/io/openliberty/guides/query/QueryResource.java`
----

QueryResource.java
[source, java, linenums, role='code_column hide_tags=copyright']
----
include::defaultrx/query/src/main/java/io/openliberty/guides/query/QueryResource.java[]
----

The [hotspot=systemLoadMethod file=1]`systemLoad` endpoint asynchronously processes the data that is retrieved by the [hotspot=getsystem file=1]`InventoryClient` and serves the data when all of the services have responded. The [hotspot=thenAcceptAsync file=1]`thenAcceptAsync()` and [hotspot=exceptionally file=1]`exceptionally()` methods together behave like an asynchronous try-catch block. The data is processed in the [hotspot=thenAcceptAsync file=1]`thenAcceptAsync()` method only when the `CompletionStage` interface has finished retrieving it. When you return a `CompletionStage` type in the resource, it doesn’t necessarily mean that the computation completed and the response was built.

A [hotspot=countdownlatch file=1]`CountDownLatch` object is used to keep track of how many asynchronous requests are being waited on. 

> The `CountDownLatch` object will countdown towards 0 after each completion of a thread via the [hotspot=countdown1 hotspot=countdown2 file=1]`countdown` method call. 

Reviewer 2: Make this present tense: "...object counts down towards `0` after..." It's describing how the technology works so always present tense.

[role="code_command hotspot file=0", subs="quotes"]
----
#Replace the `QueryResource` class.#
`query/src/main/java/io/openliberty/guides/query/QueryResource.java`
----

The goal of the [hotspot=systemload file=0]`systemLoad()` method is to return the system with the largest load and the system with the smallest load. 
The `systemLoad` endpoint first gets all of the hostnames by calling [hotspot=getSystems file=0]`getSystems()`. 
Then it loops through each one and calls [hotspot=getSystem file=0]`getSystem()` on them.

Instead of using `thenAcceptAsync()`, `Observable` uses the [hotspot=subscribe file=0]`subscribe()` method to asynchronously process data. 
Thus, any required processing of data will be performed inside the `subscribe()` method. In this case, it is simply saving the data in the temporary [hotspot=holderClass file=0]`Holder` class. 
The `Holder` class is used to store the value returned from the client because values cannot be returned inside the `subscribe()` method. 
The highest and lowest load systems are updated in the [hotspot=updateHighestMethod file=0]`updateHighest()` and [hotspot=updateLowestMethod file=0]`updateLowest()` methods respectively.

> == Re-Building and running the application

Reviewer 2: "Re-building..." (don't capitalise "building")

The JSON output has a `highest` attribute which represents the system with the highest load. Similarly, the `lowest` attribute represents the system with the lowest load. The JSON output for each of these feature the respective hostname and systemLoad of the system.

When you are done checking out the application, run the following command to stop the query microservice. We will leave the `system` and `inventory` services running because we will use them when we rebuild the application later on. 

--
[role='command']
```
docker stop query
```
--

== Updating the web client to use an alternative reactive provider

pom.xml
[source, xml, linenums, role="code_column"]
----
include::finish/query/pom.xml[]
----

> Although Open Liberty provides the Apache CXF JAX-RS reactive provider by default, you can alternatively use another provider, such as Eclipse Jersey or RESTEasy, which also support the use of https://github.com/ReactiveX/RxJava[RxJava^] reactive libraries. 

Reviewer: Same as before.
```suggestion
Although Open Liberty provides the default reactive provider which returns `CompletionStage` types, you  can alternatively use another provider that supports the use of other reactive frameworks like https://github.com/ReactiveX/RxJava[RxJava^] .  The Apache CXF and Eclipse Jersey projects produce such providers.
```

While RESTEasy likely also produces reactive providers, we have not tested them with Liberty. It is possible that RESTEasy's providers will only work with a RESTEasy JAX-RS runtime (like Wildfly or Quarkus), so that is why I don't recommend that we mention RESTEasy here.  It is certainly possible that RESTEasy's provider would work with Liberty with no problems, but I wouldn't want to advertise it until we have tested it.

> You'll now update the web client to use the Jersey reactive provider instead of CXF. 

Reviewer: ```suggestion
You'll now update the web client to use the Jersey reactive provider for RxJava. 
```

You can learn more with this post about https://openliberty.io/blog/2019/04/10/jaxrs-reactive-extensions.html[JAX-RS reactive extensions with RxJava Backpressure^]

== Updating the REST resource to support the reactive JAX-RS client

QueryResource.java
[source, java, linenums, role='code_column hide_tags=copyright']
----
include::finish/query/src/main/java/io/openliberty/guides/query/QueryResource.java[]
----

Now that the client methods return the `Observable` class, you must update the resource to accommodate these changes.

[role="code_command hotspot file=0", subs="quotes"]
----
#Replace the `QueryResource` class.#
`query/src/main/java/io/openliberty/guides/query/QueryResource.java`
----

The goal of the [hotspot=systemload file=0]`systemLoad()` method is to return the system with the largest load and the system with the smallest load. 
The `systemLoad` endpoint first gets all of the hostnames by calling [hotspot=getSystems file=0]`getSystems()`. 
Then it loops through each one and calls [hotspot=getSystem file=0]`getSystem()` on them.

Instead of using `thenAcceptAsync()`, `Observable` uses the [hotspot=subscribe file=0]`subscribe()` method to asynchronously process data. 

> Thus, any required processing of data will be performed inside the `subscribe()` method. In this case, it is simply saving the data in the temporary [hotspot=holderClass file=0]`Holder` class. 

Reviewer 2: "...is performed inside..."

== Great work! You're done!

> You have just modified an application to make HTTP requests using a reactive JAX-RS client with Open Liberty and Jersey.

Reviewer: ```suggestion
You have just modified an application to make HTTP requests using a reactive JAX-RS client with Open Liberty and Jersey's RxJava provider.
```
